### DISCUSSION TITLE
Proposal: Infer generic arguments type by pattern matching on where constructs

### DISCUSSION THREAD

**Discussion by tmm360** [dotnet/csharplang#772] [General]
I'm not sure that this had not been discussed, but from a rapid search I didn't find this proposal.
I've code like this:

    class ModelBase<TKey>
    {
        TKey Id { get; set; }
    }
    class Repo<TModel, TKey> :
        where TModel : ModelBase<TKey>
    {
        //code where I access both TModel and TKey types
    }

In my Repo class, I need to access both the `TModel` and `TKey`, but when I specific the `TModel` argument, a pattern matching system should be able to extract the `TKey` automatically.

For use the `TKey` generic type, I have to declare it as a generic argument, so every time that I need to create a Repo, I have to specific model and relative key types.

My propose is to implement generic type inference from `where` constructs, so `TKey` could be removed from class declaration, and Repo could be written like this:

    class Repo<TModel> :
        where TModel : ModelBase<TKey>
    {
        //code where I access both TModel and TKey types
    }

**Comment by HaloFour** [comment]
The problem with that approach is that the CLR treats generic types as unique based on their arity of generic type parameters, so `Repo<TModel>` and `Repo<TModel, TKey>` are technically two distinct types.  To the CLR (and any other languauge), this type would still *have* to be `Repo<TModel, TKey>`.

**Comment by tmm360** [comment]
And if Repo class declaration would be something like this?

    class Repo<TModel, implicit TKey> :
        where TModel : ModelBase<TKey>
    {
        //code where I access both TModel and TKey types
    }

This could say to CLR that type is a `Repo<TModel, TKey>`, but it doesn't force you to explicit that. Obviously, if I declare also a `Repo<TClass>` class an ambiguous warning should be raised.

For avoid ambiguity, I could also choose a keyword for autocomplete the field, or simply leave it empty like this:

    class MyModel : ModelBase<string>
    { }
    
    var repo = new Repo<MyModel, >(); //repo is Repo<MyModel, string>
